src/components/layout/ChatArea.tsx

```diff
@@ -263,6 +280,32 @@ export const ChatArea: React.FC = () => {
     inputRef.current?.focus();
   });
 
+  useEffect(() => {
+    const handleTopicUpdate = ({
+      serverId,
+      channelId,
+      channelTopic,
+    }: { serverId: string; channelId: string; channelTopic: string }) => {
+      useStore.setState((state) => {
+        const updatedServers = state.servers.map((server) => {
+          if (server.id === serverId) {
+            const updatedChannels = server.channels.map((channel) => {
+              if (channel.id === channelId) {
+                return { ...channel, topic: channelTopic };
+              }
+              return channel;
+            });
+            return { ...server, channels: updatedChannels };
+          }
+          return server;
+        });
+        return { servers: updatedServers };
+      });
+    };
+
+    ircClient.on("TOPIC", handleTopicUpdate);
+  }, []);
```

Comment:
No need to have the handler inside useEffect I would say:
```suggestion
  const handleTopicUpdate = ({
    serverId,
    channelId,
    channelTopic,
  }: { serverId: string; channelId: string; channelTopic: string }) => {
    useStore.setState((state) => {
      const updatedServers = state.servers.map((server) => {
        if (server.id === serverId) {
          const updatedChannels = server.channels.map((channel) => {
            if (channel.id === channelId) {
              return { ...channel, topic: channelTopic };
            }
            return channel;
          });
          return { ...server, channels: updatedChannels };
        }
        return server;
      });
      return { servers: updatedServers };
    });
  };
  useEffect(() => {
    ircClient.on("TOPIC", handleTopicUpdate);
  }, []);
```